File: LintCode_Python/threeSumClosest.py
# Given an array S of n integers, find three integers in S such that the sum is
# closest to a given number, target. Return the sum of the three integers.

# Have you met this question in a real interview? Yes
# Example
# For example, given array S = {-1 2 1 -4}, and target = 1. The sum that is closest
#  to the target is 2. (-1 + 2 + 1 = 2).

# Note
# You may assume that each input would have exactly one solution.

# Challenge
# O(n^2) time, O(1) extra space

import sys

def threeSumClosest(numbers, target):
    if not numbers or len(numbers) < 3:
        return None
    numbers.sort()
    n = len(numbers)
    result = sys.maxint
    for i in range(n - 2):
        left = i + 1
        right = n - 1
        while left < right:
            sums = numbers[i] + numbers[left] + numbers[right]
            if sums == target:
                return sums
            elif sums < target:
                left += 1
            else:
                right -= 1
            if abs(sums - target) < abs(result - target):
                result = sums

    return result


# print threeSumClosest([-1, 2, 1, -4], 1)
# print threeSumClosest([-2,-3,5,-1,-4,5,-11,7,1,2,3,4,-7,-1,-2,-3,-4,-5], 45)
print threeSumClosest([-2,-3,-4,-5,-100,99,1,4,4,4,5,1,0,-1,2,3,4,5], 77)

File: LeetCode_Java/src/ThreeSumClosest.java
import java.util.Arrays;

/**
 * Given an array S of n integers, find three integers in S such that the sum is closest to a given 
 * number, target. Return the sum of the three integers. You may assume that each input would have 
 * exactly one solution
 * For example, given array S = {-1 2 1 -4}, and target = 1.
 * The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
 * 
 * Tags: Array, Two pointers
 * @author wendi
 *
 */
public class ThreeSumClosest {

	/**
	 * Two pointers: 3 pointers to point current element, next element and the last element. If the 
	 * sum is less than target, it means we have to add a larger element so next element move to 
	 * the next. If the sum is greater, it means we have to add a smaller element so last element 
	 * move to the second last element. Keep doing this until the end. Each time compare the 
	 * difference between sum and target, if it is less than minimum difference so far, then replace 
	 * result with it, otherwise keep iterating.
	 * @param int[] nums, int target
	 * @return int
	 * Time: O(n^2)
	 * Space: O(1)
	 */
	public int threeSumClosest(int[] nums, int target) {
		if (nums == null || nums.length < 3) return 0;
		Arrays.sort(nums);
		int result = nums[0] + nums[1] + nums[nums.length - 1];
		for (int i = 0; i <= nums.length - 3; i++) {
			int start = i + 1;
			int end = nums.length - 1;
			while (start < end) {
				int sum = nums[i] + nums[start] + nums[end];
				if (sum == target) return sum;
				else if (sum < target) start++;
				else end--;
				if (Math.abs(target - sum) < Math.abs(target - result)) {
					result = sum;
				}
			}
		}
		return result;
	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		ThreeSumClosest result = new ThreeSumClosest();
		System.out.println(result.threeSumClosest(new int[] {-1, 2, 1, -4}, 1));
		System.out.println(result.threeSumClosest(new int[] {0, 0, 0}, 1));
	}

}
